docs/arch/adr-003-versioning-branching-strategy.md

```diff
@@ -0,0 +1,66 @@
+# Release Versioning / Branching Strategy
+
+## Context
+With `dbt` ever evolving and progressing, we continue to ship and release versions on a regular basis. As we release `dbt`, versioning and branching become key to identifying what changes for users and conveying the expectations of those changes. We currently push all our changes to `main` which is the latest version of our code. We don't always want to release the latest version of our code though. We want to be able to isolate changes to go into particular releases based on urgency (e.g. bug fixes) and stability (e.g. new features). 
+
+We already have an established versioning strategy when it comes to major/minor/patch releases. The following is our approach as to what to expect from each release version:
+
+* Major: This signifies that breaking changes which require user action are included in this release.
+* Minor version: This signifies that new features are included in this release.
```

Comment:
are we making the guarantee that there will be no breaking changes here either?

Reply:
In minor releases? Correct, no breaking changes. CC @jtcohen6 to double confirm 😄 b/c I know we did put a breaking changelog in the other day...

Reply:
No breaking changes for project code / end users of dbt!!!

We reserve the right to make breaking changes ONLY for:
- the adapter interface (Python methods), relevant to maintainers of adapter plugins
- metadata interfaces — new versions of JSON artifact versions, new version of structured logging

We should still aim to avoid those changes whenever possible. When they do happen, we should communicate them proactively and clearly, and call out _exactly_ who they're for. Probably adapter plugin maintainers. E.g.
https://github.com/dbt-labs/dbt-core/blob/9c233f27ab280f36cc5b3007d0b85a0f9d3ad13a/.changes/1.1.0/Breaking%20Changes-20220316-031311.yaml#L3
https://github.com/dbt-labs/dbt-core/blob/9c233f27ab280f36cc5b3007d0b85a0f9d3ad13a/.changes/1.1.0/Breaking%20Changes-20220412-152450.yaml#L2